File: src/main/java/com/fengsheng/skill/ChengZhi.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.card.Card;
import com.fengsheng.phase.DieSkill;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

/**
 * 顾小梦技能【承志】：一名其他角色死亡前，若此角色牌已翻开，则你获得其所有手牌，并查看其身份牌，你可以获得该身份牌，并将你原本的身份牌面朝下移出游戏。
 */
public class ChengZhi extends AbstractSkill implements TriggeredSkill {
    @Override
    public SkillId getSkillId() {
        return SkillId.CHENG_ZHI;
    }

    @Override
    public ResolveResult execute(Game g) {
        if (!(g.getFsm() instanceof DieSkill fsm) || fsm.askWhom == fsm.diedQueue.get(fsm.diedIndex) || fsm.askWhom.findSkill(getSkillId()) == null)
            return null;
        if (!fsm.askWhom.isAlive())
            return null;
        if (!fsm.askWhom.isRoleFaceUp())
            return null;
        if (fsm.askWhom.getSkillUseCount(getSkillId()) > 0)
            return null;
        fsm.askWhom.addSkillUseCount(getSkillId());
        return new ResolveResult(new executeChengZhi(fsm), true);
    }

    private record executeChengZhi(DieSkill fsm) implements WaitingFsm {
        private static final Logger log = Logger.getLogger(executeChengZhi.class);

        @Override
        public ResolveResult resolve() {
            Player r = fsm.askWhom;
            Player whoDie = fsm.diedQueue.get(fsm.diedIndex);
            Card[] cards = whoDie.deleteAllCards();
            r.addCard(cards);
            log.info(r + "发动了[承志]，获得了" + whoDie + "的" + Arrays.toString(cards));
            if (whoDie.getIdentity() == Common.color.Has_No_Identity)
                return new ResolveResult(fsm, true);
            for (Player player : r.getGame().getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    var builder = Role.skill_wait_for_cheng_zhi_toc.newBuilder();
                    builder.setPlayerId(p.getAlternativeLocation(r.location())).setWaitingSecond(20);
                    builder.setDiePlayerId(p.getAlternativeLocation(whoDie.location()));
                    if (p == r) {
                        for (Card card : cards) builder.addCards(card.toPbCard());
                        builder.setIdentity(whoDie.getIdentity());
                        builder.setSecretTask(whoDie.getSecretTask());
                        final int seq2 = p.getSeq();
                        builder.setSeq(seq2);
                        p.setTimeout(GameExecutor.post(r.getGame(), () -> r.getGame().tryContinueResolveProtocol(r, Role.skill_cheng_zhi_tos.newBuilder().setEnable(false).setSeq(seq2).build()), p.getWaitSeconds(builder.getWaitingSecond() + 2), TimeUnit.SECONDS));
                    }
                    p.send(builder.build());
                }
            }
            if (r instanceof RobotPlayer)
                GameExecutor.post(r.getGame(), () -> r.getGame().tryContinueResolveProtocol(r, Role.skill_cheng_zhi_tos.newBuilder().setEnable(true).build()), 2, TimeUnit.SECONDS);
            return null;
        }

        @Override
        public ResolveResult resolveProtocol(Player player, GeneratedMessageV3 message) {
            if (player != fsm.askWhom) {
                log.error("不是你发技能的时机");
                return null;
            }
            if (!(message instanceof Role.skill_cheng_zhi_tos pb)) {
                log.error("错误的协议");
                return null;
            }
            Player r = fsm.askWhom;
            Player whoDie = fsm.diedQueue.get(fsm.diedIndex);
            Game g = r.getGame();
            if (r instanceof HumanPlayer humanPlayer && !humanPlayer.checkSeq(pb.getSeq())) {
                log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
                return null;
            }
            if (!pb.getEnable()) {
                r.incrSeq();
                for (Player p : g.getPlayers()) {
                    if (p instanceof HumanPlayer player1)
                        player1.send(Role.skill_cheng_zhi_toc.newBuilder().setEnable(false).setPlayerId(player1.getAlternativeLocation(r.location()))
                                .setDiePlayerId(player1.getAlternativeLocation(whoDie.location())).build());
                }
                return new ResolveResult(fsm, true);
            }
            r.incrSeq();
            r.setIdentity(whoDie.getIdentity());
            r.setSecretTask(whoDie.getSecretTask());
            whoDie.setIdentity(Common.color.Has_No_Identity);
            log.info(r + "获得了" + whoDie + "的身份牌");
            for (Player p : g.getPlayers()) {
                if (p instanceof HumanPlayer player1)
                    player1.send(Role.skill_cheng_zhi_toc.newBuilder().setEnable(true).setPlayerId(player1.getAlternativeLocation(r.location()))
                            .setDiePlayerId(player1.getAlternativeLocation(whoDie.location())).build());
            }
            return new ResolveResult(fsm, true);
        }
    }
}

File: src/main/kotlin/skill/ChengZhi.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.card.count
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.color.*
import com.fengsheng.protos.Common.secret_task
import com.fengsheng.protos.Common.secret_task.*
import com.fengsheng.protos.Role.skill_cheng_zhi_tos
import com.fengsheng.protos.skillChengZhiToc
import com.fengsheng.protos.skillChengZhiTos
import com.fengsheng.protos.skillWaitForChengZhiToc
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 顾小梦技能【承志】：一名其他角色死亡前，若此角色牌已翻开，则你获得其所有手牌，并查看其身份牌，你可以获得该身份牌，并将你原本的身份牌面朝下移出游戏。
 */
class ChengZhi : TriggeredSkill {
    override val skillId = SkillId.CHENG_ZHI

    override val isInitialSkill = true

    override fun execute(g: Game, askWhom: Player): ResolveResult? {
        val event = g.findEvent<PlayerDieEvent>(this) { event ->
            askWhom !== event.whoDie || return@findEvent false
            askWhom.alive || return@findEvent false
            askWhom.roleFaceUp
        } ?: return null
        val whoDie = event.whoDie
        return ResolveResult(
            ExecuteChengZhi(g.fsm!!, askWhom, whoDie, whoDie.cards.isNotEmpty()),
            true
        )
    }

    private data class ExecuteChengZhi(
        val fsm: Fsm,
        val r: Player,
        val whoDie: Player,
        val hasCard: Boolean
    ) : WaitingFsm {
        override val whoseTurn: Player
            get() = fsm.whoseTurn

        override fun resolve(): ResolveResult? {
            val cards = whoDie.cards.toList()
            whoDie.cards.clear()
            r.cards.addAll(cards)
            if (hasCard) {
                logger.info("${r}发动了[承志]，获得了${whoDie}的${cards.joinToString()}并查看身份牌")
                r.game!!.addEvent(GiveCardEvent(whoseTurn, whoDie, r))
            } else {
                logger.info("${r}发动了[承志]，查看了${whoDie}的身份牌")
            }
            if (whoDie.identity == Has_No_Identity) return ResolveResult(fsm, true)
            r.game!!.players.send { player ->
                skillWaitForChengZhiToc {
                    playerId = player.getAlternativeLocation(r.location)
                    waitingSecond = Config.WaitSecond
                    diePlayerId = player.getAlternativeLocation(whoDie.location)
                    if (player === r) {
                        cards.forEach { this.cards.add(it.toPbCard()) }
                        identity = whoDie.identity
                        secretTask = whoDie.secretTask
                        val seq2 = player.seq
                        seq = seq2
                        player.timeout = GameExecutor.post(r.game!!, {
                            if (r.checkSeq(seq2)) {
                                r.game!!.tryContinueResolveProtocol(r, skillChengZhiTos {
                                    enable = false
                                    seq = seq2
                                })
                            }
                        }, player.getWaitSeconds(waitingSecond + 2).toLong(), TimeUnit.SECONDS)
                    }
                }
            }
            if (r is RobotPlayer) GameExecutor.post(r.game!!, {
                fun Player.process(identity: color, secretTask: secret_task) = if (identity != Black) {
                    maxOf(game!!.players.filter { it!!.alive && it.identity == identity }
                        .maxOfOrNull { it!!.messageCards.count(identity) * 10 } ?: 0,
                        r.messageCards.count(identity) * 10
                    )
                } else when (secretTask) {
                    Collector -> maxOf(messageCards.count(Red), messageCards.count(Blue)) * 2 - 3
                    Mutator -> game!!.players.filter { it!!.alive }
                        .maxOf { maxOf(it!!.messageCards.count(Red), it.messageCards.count(Blue)) * 2 - 2 }

                    Pioneer -> messageCards.count(Black) * 10 - 1
                    else -> -100
                }
                r.game!!.tryContinueResolveProtocol(r, skillChengZhiTos {
                    enable = whoDie.identity == r.identity && whoDie.secretTask == r.secretTask ||
                        r.process(whoDie.identity, whoDie.secretTask) > r.process(r.identity, r.secretTask)
                })
            }, 3, TimeUnit.SECONDS)
            return null
        }

        override fun resolveProtocol(player: Player, message: GeneratedMessage): ResolveResult? {
            if (player !== r) {
                logger.error("不是你发技能的时机")
                player.sendErrorMessage("不是你发技能的时机")
                return null
            }
            if (message !is skill_cheng_zhi_tos) {
                logger.error("错误的协议")
                player.sendErrorMessage("错误的协议")
                return null
            }
            val g = r.game!!
            if (r is HumanPlayer && !r.checkSeq(message.seq)) {
                logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${message.seq}")
                r.sendErrorMessage("操作太晚了")
                return null
            }
            if (!message.enable) {
                r.incrSeq()
                g.players.send {
                    skillChengZhiToc {
                        enable = false
                        playerId = it.getAlternativeLocation(r.location)
                        diePlayerId = it.getAlternativeLocation(whoDie.location)
                    }
                }
                return ResolveResult(fsm, true)
            }
            r.incrSeq()
            r.identity = whoDie.identity
            r.secretTask = whoDie.secretTask
            whoDie.identity = Has_No_Identity
            logger.info("${r}获得了${whoDie}的身份牌")
            g.players.send {
                skillChengZhiToc {
                    enable = true
                    playerId = it.getAlternativeLocation(r.location)
                    diePlayerId = it.getAlternativeLocation(whoDie.location)
                }
            }
            return ResolveResult(fsm, true)
        }
    }
}
